### ISSUE TITLE
Error in Oracle when adding an element to a certificate

### ISSUE THREAD

**Issue by markn86** [markn86/moodle-mod_customcert#123]
Reported by Iván Sibillà.

```
Debug info: ORA-01747: especificación no válida de usuario.tabla.columna,
tabla.columna o columna
INSERT INTO m_customcert_elements
(name,data,font,size,colour,posx,posy,width,refpoint,timemodified,element,pageid,sequence,timecreated)
VALUES
(:name,:data,:font,:size,:colour,:posx,:posy,:width,:refpoint,:timemodified,:element,:pageid,:sequence,:timecreated)
[array (
'name' => 'Logo',
'data' =>
'{"contextid":"1","filearea":"image","itemid":"0","filepath":"\\/","filename":"Logo_Trans.png","width":80,"height":0}',
'font' => NULL,
'size' => NULL,
'colour' => NULL,
'posx' => 65,
'posy' => 1,
'width' => 80,
'refpoint' => NULL,
'timemodified' => 1500879365,
'element' => 'image',
'pageid' => 21,
'sequence' => 1,
'timecreated' => 1500879365,
)]
Error code: dmlwriteexception
Stack trace:

    line 477 of /lib/dml/moodle_database.php: dml_write_exception thrown
    line 271 of /lib/dml/oci_native_moodle_database.php: call to
moodle_database->query_end()
    line 1284 of /lib/dml/oci_native_moodle_database.php: call to
oci_native_moodle_database->query_end()
    line 1331 of /lib/dml/oci_native_moodle_database.php: call to
oci_native_moodle_database->insert_record_raw()
    line 151 of /mod/customcert/classes/element.php: call to
oci_native_moodle_database->insert_record()
    line 135 of /mod/customcert/element/image/classes/element.php: call to
mod_customcert\element->save_form_elements()
    line 90 of /mod/customcert/edit_element.php: call to
customcertelement_image\element->save_form_elements()
```

**Comment by markn86** [comment]
After eventually setting up a Moodle site running Oracle I was able to replicate this issue.

**Comment by scara** [comment]
Hi @markn86,
`size` is a reserved word in Oracle: https://github.com/moodle/moodle/blob/b3e2aaebc39653f94143dfa6a542fb945d3989ca/lib/ddl/oracle_sql_generator.php#L697 .

See also: https://tracker.moodle.org/browse/MDLSITE-5170 and https://github.com/scara/moodle/compare/f4a2d69...m34_WIP_Properly_Escape_Column_Names_Reserved_Words_Proof.

HTH,
Matteo

**Comment by markn86** [comment]
Hi Matteo,

Thanks for the information.

Looks like I will have to change the column, probably to something like 'fontsize'. I also wonder if this is a reserved word on any of the other supported databases I never use (mssql with all the wierd combinations of sqlsrv and freetds).

Your patch looks like it is going in the right direction but in the meantime the approach here would be to rename the column.

**Comment by markn86** [comment]
![selection_028](https://user-images.githubusercontent.com/783213/28554917-fcc56cfc-712e-11e7-9e58-ee0c7f814ef5.png)

Doh, if only I knew about the admin/tool/xmldb check for reserved words

**Comment by scara** [comment]
Hi @markn86,
I think that those check could be sensitive i.e. vote for the precheck (MDLSITE-5170), in case.

The proposal in Moodle should be voted someway and I thought to pass through the developer forum instead of the Tracker... but I didn't gain the expected attention there.

**Comment by scara** [comment]
@markn86: TNX for voting, appreciated 😄!

**Comment by markn86** [comment]
This issue will be the next one I work on when I have time as it makes the plugin unusable in Oracle.

**Comment by scara** [comment]
You could _escape_ it when the DB is Oracle, if you want to keep that column name i.e. you prefer not to migrate the contents.

HTH,
Matteo

**Comment by scara** [comment]
FYI: https://tracker.moodle.org/browse/MDL-59635.
If that will be integrated, you could use `sql_column_name_escape()`.

HTH,
Matteo

**Comment by markn86** [comment]
Thanks Matteo, I did consider doing this but I think the proper solution is to avoid using reserved words, they are reserved for a reason. This will prevent future mishaps as well and isn't that difficult to do.

**Comment by markn86** [comment]
Created fix for 3.3 and 3.2 (versions I am currently supporting). Closing.
